### PR TITLE
Improve target discovery pipeline

### PR DESCRIPTION
**Pull request by fabxc**
Replace the TargetProvider Stop method with done channels
that ensure properly broadcasted shutdown of the whole pipeline.

This is without a doubt a fairly invasive change into a component that
has always been problematic. So I would like to get this into master as
soon as possible so we get timely feedback.
There is a prerequisite for further changes.

@juliusv

### CODE REVIEW

File: retrieval/discovery/consul.go [hunk 1]
```diff
@@ -133,18 +133,22 @@ func (cd *ConsulDiscovery) Sources() []string {
...
 	defer close(ch)
+	defer cd.stop()
 
 	update := make(chan *consulService, 10)
-	go cd.watchServices(update)
+	go cd.watchServices(update, done)
 
 	for {
 		select {
-		case <-cd.runDone:
```
Comment (juliusv) [inline retrieval/discovery/consul.go]: This channel is now unused, but the definition/make is still there.

File: retrieval/targetmanager.go [hunk 2]
```diff
@@ -107,24 +170,32 @@ func (tm *TargetManager) Run() {
...
 // handleTargetUpdates receives target group updates and handles them in the
 // context of the given job config.
-func (tm *TargetManager) handleTargetUpdates(cfg *config.ScrapeConfig, ch <-chan *config.TargetGroup) {
-	for tg := range ch {
-		log.Debugf("Received potential update for target group %q", tg.Source)
+func (tm *TargetManager) handleUpdates(ch <-chan targetGroupUpdate, done <-chan struct{}) {
+	for {
+		select {
+		case update := <-ch:
+			if update.tg == nil {
```
Comment (juliusv) [inline retrieval/targetmanager.go]: You already check for `nil` in the sender - do we need a check in both places?

File: retrieval/discovery/consul.go [hunk 3]
```diff
@@ -191,7 +184,7 @@ func (cd *ConsulDiscovery) watchServices(update chan<- *consulService) {
 		})
 		if err != nil {
 			log.Errorf("Error refreshing service list: %s", err)
-			<-time.After(consulRetryInterval)
+			time.Sleep(consulRetryInterval)
```
Comment (juliusv) [inline retrieval/discovery/consul.go]: How about also returning here when `done` is closed via a `select`? That way it doesn't take 15 seconds to shut down the target provider in the unlucky case.

File: retrieval/discovery/consul.go [hunk 4]
```diff
@@ -253,7 +245,7 @@ func (cd *ConsulDiscovery) watchService(srv *consulService, ch chan<- *config.Ta
 		})
 		if err != nil {
 			log.Errorf("Error refreshing service %s: %s", srv.name, err)
-			<-time.After(consulRetryInterval)
+			time.Sleep(consulRetryInterval)
```
Comment (juliusv) [inline retrieval/discovery/consul.go]: Dito

File: retrieval/discovery/file.go [hunk 5]
```diff
@@ -198,11 +197,10 @@ func fileSource(filename string, i int) string {
...
 }
 
-// Stop implements the TargetProvider interface.
-func (fd *FileDiscovery) Stop() {
+// stop shuts down the file watcher.
+func (fd *FileDiscovery) stop() {
 	log.Debugf("Stopping file discovery for %s...", fd.paths)
 
-	fd.done <- struct{}{}
 	// Closing the watcher will deadlock unless all events and errors are drained.
```
Comment (juliusv) [inline retrieval/discovery/file.go]: Ugh, unrelated, but didn't notice this before - does this mean that there's a race condition between the drain here and the closing below (if new events happen after draining, but just before closing, it deadlocks)? Doesn't seem like there's a way to fix it though?

File: retrieval/discovery/dns.go [hunk 6]
```diff
@@ -110,12 +113,6 @@ func (dd *DNSDiscovery) Run(ch chan<- *config.TargetGroup) {
 
 // Stop implements the TargetProvider interface.
 func (dd *DNSDiscovery) Stop() {
```
Comment (juliusv) [inline retrieval/discovery/dns.go]: This method can be removed now, as it's empty and no longer part of the `TargetProvider` interface.

File: retrieval/targetmanager.go [hunk 7]
```diff
@@ -40,20 +40,18 @@ type TargetProvider interface {
 	// Run hands a channel to the target provider through which it can send
 	// updated target groups. The channel must be closed by the target provider
 	// if no more updates will be sent.
```
Comment (juliusv) [inline retrieval/targetmanager.go]: Maybe mention the new argument here too?

**Comment by juliusv** [comment]
:+1: besides comments. Cool! Let's give this a good long run on a big test Prometheus soon please :)

File: retrieval/targetmanager.go (hunk 2 above)
Comment (fabxc) [inline retrieval/targetmanager.go]: Can't hurt, really.

File: retrieval/discovery/consul.go (hunk 3 above)
Comment (fabxc) [inline retrieval/discovery/consul.go]: Let's not pile even more on this PR. Same for the file stuff down there.
The whole retrieval package needs another pass so we will get to it later.

Comment (jimmidyson) [inline retrieval/discovery/consul.go]: Not for this PR but can probably reuse the `until` in kubernetes discovery to shut down target providers nicely (https://github.com/prometheus/prometheus/blob/master/retrieval/discovery/kubernetes/discovery.go#L538-L555). Borrowed from kubernetes codebase btw ;) Worth making it a common util?

Comment (fabxc) [inline retrieval/discovery/consul.go]: Sure :) we can drag it out the first time we need it again.

**Comment by jimmidyson** [comment]
Tested Kubernetes updates (briefly) & all looks good.

Although on shutdown get loads of repeated:

    time="2015-08-14T12:24:20Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175

File: retrieval/targetmanager.go (hunk 7 above)
Comment (fabxc) [inline retrieval/targetmanager.go]: Done.

**Comment by fabxc** [comment]
Cool, thanks.
How certain are you this related to this PR?

**Comment by jimmidyson** [comment]
Not at all certain. Don't remember seeing it before though.

**Comment by fabxc** [comment]
The function can basically only be called on requesting `POST /-/reload` or sending `SIGHUP`.

File: retrieval/discovery/consul.go (hunk 1 above)
Comment (fabxc) [inline retrieval/discovery/consul.go]: Removed

**Comment by jimmidyson** [comment]
ctrl-c causing this for me. Running in docker container if that makes any difference (don't think it does). Here's a log:

```
time="2015-08-14T14:20:45Z" level=warning msg="Received SIGTERM, exiting gracefully..." file=main.go line=157 
time="2015-08-14T14:20:45Z" level=info msg="See you next time!" file=main.go line=164 
time="2015-08-14T14:20:45Z" level=info msg="Stopping target manager..." file=targetmanager.go line=208 
time="2015-08-14T14:20:45Z" level=info msg="Target manager stopped." file=targetmanager.go line=221 
time="2015-08-14T14:20:45Z" level=info msg="Stopping notification handler..." file=notification.go line=214 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175 
time="2015-08-14T14:20:45Z" level=info msg="Loading configuration file /etc/prometheus/prometheus.yml" file=main.go line=175
```

**Comment by fabxc** [comment]
Oh yes, that's a regression: #988

**Comment by fabxc** [comment]
Okay, addressed comments except for those unrelated to this PR's changes.
Let's merge this so we get some feedback.

**Pull request merged**